File: aws-lambda-golang-docker-starter/internal/events/events.go
package events

import (
	"encoding/json"
	"strings"

	awsevents "github.com/aws/aws-lambda-go/events"
)

func ParseEnvelope(raw json.RawMessage) (map[string]any, error) {
	var envelope map[string]any
	err := json.Unmarshal(raw, &envelope)
	return envelope, err
}

func unmarshalAs[T any](raw json.RawMessage, match bool) (T, bool) {
	var event T
	if !match {
		return event, false
	}
	if err := json.Unmarshal(raw, &event); err != nil {
		return event, false
	}
	return event, true
}

func matchRaw(raw json.RawMessage, match bool) (json.RawMessage, bool) {
	if !match {
		return nil, false
	}
	return raw, true
}

func IsHTTPV2(raw json.RawMessage, envelope map[string]any) (awsevents.APIGatewayV2HTTPRequest, bool) {
	version, _ := envelope["version"].(string)
	if version != "2.0" {
		return awsevents.APIGatewayV2HTTPRequest{}, false
	}
	httpCtx := nestedObject(envelope, "requestContext", "http")
	if httpCtx == nil {
		return awsevents.APIGatewayV2HTTPRequest{}, false
	}
	method, _ := httpCtx["method"].(string)
	return unmarshalAs[awsevents.APIGatewayV2HTTPRequest](raw, strings.TrimSpace(method) != "")
}

func IsLambdaFunctionURL(raw json.RawMessage, envelope map[string]any) (awsevents.APIGatewayV2HTTPRequest, bool) {
	event, ok := IsHTTPV2(raw, envelope)
	if !ok {
		return event, false
	}
	rc := nestedObject(envelope, "requestContext")
	if rc == nil {
		return awsevents.APIGatewayV2HTTPRequest{}, false
	}
	domainName, _ := rc["domainName"].(string)
	if !strings.Contains(domainName, "lambda-url") {
		return awsevents.APIGatewayV2HTTPRequest{}, false
	}
	return event, true
}

func IsALB(raw json.RawMessage, envelope map[string]any) (awsevents.ALBTargetGroupRequest, bool) {
	return unmarshalAs[awsevents.ALBTargetGroupRequest](raw, nestedObject(envelope, "requestContext", "elb") != nil)
}

func IsHTTPV1(raw json.RawMessage, envelope map[string]any) (awsevents.APIGatewayProxyRequest, bool) {
	method, _ := envelope["httpMethod"].(string)
	return unmarshalAs[awsevents.APIGatewayProxyRequest](raw, strings.TrimSpace(method) != "")
}

func IsSQS(raw json.RawMessage, envelope map[string]any) (awsevents.SQSEvent, bool) {
	return unmarshalAs[awsevents.SQSEvent](raw, firstRecordField(envelope, "eventSource") == "aws:sqs")
}

func IsSNS(raw json.RawMessage, envelope map[string]any) (awsevents.SNSEvent, bool) {
	return unmarshalAs[awsevents.SNSEvent](raw, firstRecordField(envelope, "EventSource") == "aws:sns")
}

func IsS3(raw json.RawMessage, envelope map[string]any) (awsevents.S3Event, bool) {
	return unmarshalAs[awsevents.S3Event](raw, firstRecordField(envelope, "eventSource") == "aws:s3")
}

func IsSES(raw json.RawMessage, envelope map[string]any) (awsevents.SimpleEmailEvent, bool) {
	return unmarshalAs[awsevents.SimpleEmailEvent](raw, firstRecordField(envelope, "eventSource") == "aws:ses")
}

func IsDynamoDBStream(raw json.RawMessage, envelope map[string]any) (awsevents.DynamoDBEvent, bool) {
	return unmarshalAs[awsevents.DynamoDBEvent](raw, firstRecordField(envelope, "eventSource") == "aws:dynamodb")
}

func IsKinesisStream(raw json.RawMessage, envelope map[string]any) (awsevents.KinesisEvent, bool) {
	return unmarshalAs[awsevents.KinesisEvent](raw, firstRecordField(envelope, "eventSource") == "aws:kinesis")
}

func IsCloudWatchLogs(raw json.RawMessage, envelope map[string]any) (awsevents.CloudwatchLogsEvent, bool) {
	logs := nestedObject(envelope, "awslogs")
	if logs == nil {
		return awsevents.CloudwatchLogsEvent{}, false
	}
	data, _ := logs["data"].(string)
	return unmarshalAs[awsevents.CloudwatchLogsEvent](raw, strings.TrimSpace(data) != "")
}

func IsScheduledEvent(raw json.RawMessage, envelope map[string]any) (awsevents.CloudWatchEvent, bool) {
	source, _ := envelope["source"].(string)
	detailType, _ := envelope["detail-type"].(string)
	return unmarshalAs[awsevents.CloudWatchEvent](raw, source == "aws.events" && detailType == "Scheduled Event")
}

func IsCodeBuildState(raw json.RawMessage, envelope map[string]any) (awsevents.CodeBuildEvent, bool) {
	source, _ := envelope["source"].(string)
	detailType, _ := envelope["detail-type"].(string)
	return unmarshalAs[awsevents.CodeBuildEvent](raw, source == "aws.codebuild" && detailType == "CodeBuild Build State Change")
}

func IsCodeCommit(raw json.RawMessage, envelope map[string]any) (awsevents.CodeCommitEvent, bool) {
	first := firstRecord(envelope)
	if first == nil {
		return awsevents.CodeCommitEvent{}, false
	}
	_, ok := first["codecommit"].(map[string]any)
	return unmarshalAs[awsevents.CodeCommitEvent](raw, ok)
}

func IsCloudWatchAlarm(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	_, hasArn := envelope["alarmArn"].(string)
	_, hasData := envelope["alarmData"].(map[string]any)
	return matchRaw(raw, hasArn && hasData)
}

func IsConnectContactFlow(raw json.RawMessage, envelope map[string]any) (awsevents.ConnectEvent, bool) {
	name, _ := envelope["Name"].(string)
	return unmarshalAs[awsevents.ConnectEvent](raw, name == "ContactFlowEvent" && nestedObject(envelope, "Details", "ContactData") != nil)
}

func IsCloudFrontRequest(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	cf := firstRecordCF(envelope)
	if cf == nil {
		return nil, false
	}
	_, hasReq := cf["request"].(map[string]any)
	_, hasResp := cf["response"].(map[string]any)
	return matchRaw(raw, hasReq && !hasResp)
}

func IsCloudFrontResponse(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	cf := firstRecordCF(envelope)
	if cf == nil {
		return nil, false
	}
	_, hasResp := cf["response"].(map[string]any)
	return matchRaw(raw, hasResp)
}

func IsGuardDuty(raw json.RawMessage, envelope map[string]any) (awsevents.CloudWatchEvent, bool) {
	source, _ := envelope["source"].(string)
	return unmarshalAs[awsevents.CloudWatchEvent](raw, source == "aws.guardduty")
}

func IsS3Batch(raw json.RawMessage, envelope map[string]any) (awsevents.S3BatchJobEvent, bool) {
	_, hasSchema := envelope["invocationSchemaVersion"].(string)
	_, hasTasks := envelope["tasks"].([]any)
	return unmarshalAs[awsevents.S3BatchJobEvent](raw, hasSchema && hasTasks)
}

func IsS3Notification(raw json.RawMessage, envelope map[string]any) (awsevents.CloudWatchEvent, bool) {
	source, _ := envelope["source"].(string)
	detailType, _ := envelope["detail-type"].(string)
	return unmarshalAs[awsevents.CloudWatchEvent](raw, source == "aws.s3" && strings.TrimSpace(detailType) != "")
}

func IsLex(raw json.RawMessage, envelope map[string]any) (awsevents.LexEvent, bool) {
	_, hasMsgVer := envelope["messageVersion"].(string)
	_, hasIntent := envelope["currentIntent"].(map[string]any)
	_, hasBot := envelope["bot"].(map[string]any)
	return unmarshalAs[awsevents.LexEvent](raw, hasMsgVer && hasIntent && hasBot)
}

func IsLexV2(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	_, hasMsgVer := envelope["messageVersion"].(string)
	_, hasBot := envelope["bot"].(map[string]any)
	_, hasSession := envelope["sessionId"].(string)
	return matchRaw(raw, hasMsgVer && hasBot && hasSession)
}

func IsAppSyncResolver(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	info := nestedObject(envelope, "info")
	if info == nil {
		return nil, false
	}
	_, hasField := info["fieldName"].(string)
	_, hasReq := envelope["request"].(map[string]any)
	return matchRaw(raw, hasField && hasReq)
}

func IsAmplifyResolver(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	_, hasType := envelope["typeName"].(string)
	_, hasField := envelope["fieldName"].(string)
	_, hasReq := envelope["request"].(map[string]any)
	return matchRaw(raw, hasType && hasField && hasReq)
}

func IsAPIGatewayAuthorizer(raw json.RawMessage, envelope map[string]any) (awsevents.APIGatewayCustomAuthorizerRequest, bool) {
	eventType, ok := envelope["type"].(string)
	if !ok || (eventType != "TOKEN" && eventType != "REQUEST") {
		return awsevents.APIGatewayCustomAuthorizerRequest{}, false
	}
	_, hasMethod := envelope["methodArn"].(string)
	_, hasRoute := envelope["routeArn"].(string)
	return unmarshalAs[awsevents.APIGatewayCustomAuthorizerRequest](raw, hasMethod || hasRoute)
}

func IsAutoScalingScaleIn(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	_, hasARN := envelope["AutoScalingGroupARN"].(string)
	_, hasCap := envelope["CapacityToTerminate"].([]any)
	return matchRaw(raw, hasARN && hasCap)
}

func IsCloudFormationCustomResource(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	for _, key := range []string{"RequestType", "ResponseURL", "StackId", "RequestId", "LogicalResourceId", "ResourceType"} {
		if _, ok := envelope[key].(string); !ok {
			return nil, false
		}
	}
	return raw, true
}

func IsIoTCustomAuthorizer(raw json.RawMessage, envelope map[string]any) (awsevents.IoTCoreCustomAuthorizerRequest, bool) {
	_, hasProto := envelope["protocolData"].(map[string]any)
	_, hasConn := envelope["connectionMetadata"].(map[string]any)
	return unmarshalAs[awsevents.IoTCoreCustomAuthorizerRequest](raw, hasProto && hasConn)
}

func IsEventBridge(raw json.RawMessage, envelope map[string]any) (awsevents.CloudWatchEvent, bool) {
	source, _ := envelope["source"].(string)
	detailType, _ := envelope["detail-type"].(string)
	return unmarshalAs[awsevents.CloudWatchEvent](raw, strings.TrimSpace(source) != "" && strings.TrimSpace(detailType) != "")
}

func IsCodePipeline(raw json.RawMessage, envelope map[string]any) (awsevents.CodePipelineEvent, bool) {
	job := nestedObject(envelope, "CodePipeline.job")
	if job == nil {
		return awsevents.CodePipelineEvent{}, false
	}
	id, _ := job["id"].(string)
	return unmarshalAs[awsevents.CodePipelineEvent](raw, strings.TrimSpace(id) != "")
}

func IsFirehoseTransformation(raw json.RawMessage, envelope map[string]any) (awsevents.KinesisFirehoseEvent, bool) {
	invocationID, _ := envelope["invocationId"].(string)
	records, _ := envelope["records"].([]any)
	return unmarshalAs[awsevents.KinesisFirehoseEvent](raw, strings.TrimSpace(invocationID) != "" && len(records) > 0)
}

func IsSecretsManagerRotation(raw json.RawMessage, envelope map[string]any) (awsevents.SecretsManagerSecretRotationEvent, bool) {
	step, _ := envelope["Step"].(string)
	secretID, _ := envelope["SecretId"].(string)
	token, _ := envelope["ClientRequestToken"].(string)
	return unmarshalAs[awsevents.SecretsManagerSecretRotationEvent](raw,
		strings.TrimSpace(step) != "" && strings.TrimSpace(secretID) != "" && strings.TrimSpace(token) != "")
}

func IsTransferFamilyAuth(raw json.RawMessage, envelope map[string]any) (json.RawMessage, bool) {
	_, hasUser := envelope["username"].(string)
	_, hasServer := envelope["serverId"].(string)
	_, hasIP := envelope["sourceIp"].(string)
	return matchRaw(raw, hasUser && hasServer && hasIP)
}

func IsMSK(raw json.RawMessage, envelope map[string]any) (awsevents.KafkaEvent, bool) {
	source, _ := envelope["eventSource"].(string)
	records, _ := envelope["records"].(map[string]any)
	return unmarshalAs[awsevents.KafkaEvent](raw, source == "aws:kafka" && len(records) > 0)
}

func IsSelfManagedKafka(raw json.RawMessage, envelope map[string]any) (awsevents.KafkaEvent, bool) {
	source, _ := envelope["eventSource"].(string)
	records, _ := envelope["records"].(map[string]any)
	return unmarshalAs[awsevents.KafkaEvent](raw, source == "SelfManagedKafka" && len(records) > 0)
}

// --- helpers ---

func nestedObject(root map[string]any, keys ...string) map[string]any {
	current := root
	for _, key := range keys {
		next, ok := current[key].(map[string]any)
		if !ok {
			return nil
		}
		current = next
	}
	return current
}

func firstRecord(envelope map[string]any) map[string]any {
	records, ok := envelope["Records"].([]any)
	if !ok || len(records) == 0 {
		return nil
	}
	first, _ := records[0].(map[string]any)
	return first
}

func firstRecordField(envelope map[string]any, key string) string {
	rec := firstRecord(envelope)
	if rec == nil {
		return ""
	}
	val, _ := rec[key].(string)
	return val
}

func firstRecordCF(envelope map[string]any) map[string]any {
	rec := firstRecord(envelope)
	if rec == nil {
		return nil
	}
	cf, _ := rec["cf"].(map[string]any)
	return cf
}

File: aws-lambda-python-docker-starter/app/events/events.py
from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import Any

from aws_lambda_typing import events as aws_events

try:
    from typing import TypeGuard
except ImportError:
    from typing_extensions import TypeGuard


def _is_object(value: Any) -> TypeGuard[Mapping[str, Any]]:
    return isinstance(value, dict)


def _has_string(obj: Mapping[str, Any], key: str) -> bool:
    return isinstance(obj.get(key), str)


def _nested_object(root: Mapping[str, Any], *keys: str) -> Mapping[str, Any] | None:
    current: Any = root
    for key in keys:
        if not _is_object(current):
            return None
        current = current.get(key)
    return current if _is_object(current) else None


def _first_record(envelope: Mapping[str, Any]) -> Mapping[str, Any] | None:
    records = envelope.get("Records")
    if not isinstance(records, list) or not records:
        return None
    first = records[0]
    return first if _is_object(first) else None


def _first_record_field(envelope: Mapping[str, Any], key: str) -> str:
    rec = _first_record(envelope)
    if rec is None:
        return ""
    val = rec.get(key)
    return val if isinstance(val, str) else ""


def _first_record_cf(envelope: Mapping[str, Any]) -> Mapping[str, Any] | None:
    rec = _first_record(envelope)
    if rec is None:
        return None
    cf = rec.get("cf")
    return cf if _is_object(cf) else None


# --- HTTP-family events ---

def is_http_v1(event: Any) -> TypeGuard[aws_events.APIGatewayProxyEventV1]:
    return _is_object(event) and isinstance(event.get("httpMethod"), str)


def is_http_v2(event: Any) -> TypeGuard[aws_events.APIGatewayProxyEventV2]:
    if not _is_object(event) or event.get("version") != "2.0":
        return False
    http = _nested_object(event, "requestContext", "http")
    return http is not None and isinstance(http.get("method"), str)


def is_lambda_function_url(event: Any) -> bool:
    if not is_http_v2(event):
        return False
    rc = _nested_object(event, "requestContext")
    if rc is None:
        return False
    domain = rc.get("domainName", "")
    return isinstance(domain, str) and "lambda-url" in domain


def is_alb(event: Any) -> TypeGuard[aws_events.ALBEvent]:
    return _is_object(event) and _nested_object(event, "requestContext", "elb") is not None


def is_http_event(event: Any) -> bool:
    return is_alb(event) or is_http_v1(event) or is_http_v2(event)


# --- Queue / stream / storage events ---

def is_sqs(event: Any) -> TypeGuard[aws_events.SQSEvent]:
    return _first_record_field(event, "eventSource") == "aws:sqs" if _is_object(event) else False


def is_sns(event: Any) -> TypeGuard[aws_events.SNSEvent]:
    if not _is_object(event):
        return False
    return _first_record_field(event, "EventSource") == "aws:sns"


def is_s3(event: Any) -> TypeGuard[aws_events.S3Event]:
    if not _is_object(event):
        return False
    return _first_record_field(event, "eventSource") == "aws:s3"


def is_ses(event: Any) -> TypeGuard[aws_events.SESEvent]:
    if not _is_object(event):
        return False
    return _first_record_field(event, "eventSource") == "aws:ses"


def is_dynamodb_stream(event: Any) -> TypeGuard[aws_events.DynamoDBStreamEvent]:
    if not _is_object(event):
        return False
    return _first_record_field(event, "eventSource") == "aws:dynamodb"


def is_kinesis_stream(event: Any) -> TypeGuard[aws_events.KinesisStreamEvent]:
    if not _is_object(event):
        return False
    return _first_record_field(event, "eventSource") == "aws:kinesis"


def is_msk(event: Any) -> TypeGuard[aws_events.MSKEvent]:
    return _is_object(event) and event.get("eventSource") == "aws:kafka" and _is_object(event.get("records"))


def is_self_managed_kafka(event: Any) -> bool:
    return _is_object(event) and event.get("eventSource") == "SelfManagedKafka" and _is_object(event.get("records"))


# --- CloudWatch / EventBridge ---

def is_cloudwatch_logs(event: Any) -> TypeGuard[aws_events.CloudWatchLogsEvent]:
    if not _is_object(event):
        return False
    logs = _nested_object(event, "awslogs")
    return logs is not None and isinstance(logs.get("data"), str) and logs["data"].strip() != ""


def is_scheduled_event(event: Any) -> bool:
    return (
        _is_object(event)
        and event.get("source") == "aws.events"
        and event.get("detail-type") == "Scheduled Event"
    )


def is_codebuild_state(event: Any) -> bool:
    return (
        _is_object(event)
        and event.get("source") == "aws.codebuild"
        and event.get("detail-type") == "CodeBuild Build State Change"
    )


def is_cloudwatch_alarm(event: Any) -> bool:
    return _is_object(event) and _has_string(event, "alarmArn") and _is_object(event.get("alarmData"))


def is_eventbridge(event: Any) -> TypeGuard[aws_events.EventBridgeEvent]:
    return (
        _is_object(event)
        and isinstance(event.get("source"), str)
        and event["source"].strip() != ""
        and isinstance(event.get("detail-type"), str)
        and event["detail-type"].strip() != ""
    )


def create_eventbridge_matcher(
    source: str | None = None, detail_type: str | None = None
) -> Callable[[Any], bool]:
    def matcher(event: Any) -> bool:
        if not is_eventbridge(event):
            return False
        if source is not None and event.get("source") != source:
            return False
        if detail_type is not None and event.get("detail-type") != detail_type:
            return False
        return True

    return matcher


# --- Code* events ---

def is_codecommit(event: Any) -> TypeGuard[aws_events.CodeCommitMessageEvent]:
    if not _is_object(event):
        return False
    first = _first_record(event)
    return first is not None and _is_object(first.get("codecommit"))


def is_codepipeline(event: Any) -> TypeGuard[aws_events.CodePipelineEvent]:
    if not _is_object(event):
        return False
    job = _nested_object(event, "CodePipeline.job")
    return job is not None and isinstance(job.get("id"), str) and job["id"].strip() != ""


# --- CloudFront ---

def is_cloudfront_request(event: Any) -> bool:
    if not _is_object(event):
        return False
    cf = _first_record_cf(event)
    if cf is None:
        return False
    return _is_object(cf.get("request")) and not _is_object(cf.get("response"))


def is_cloudfront_response(event: Any) -> bool:
    if not _is_object(event):
        return False
    cf = _first_record_cf(event)
    return cf is not None and _is_object(cf.get("response"))


# --- Security ---

def is_guardduty(event: Any) -> bool:
    return _is_object(event) and event.get("source") == "aws.guardduty"


def is_connect_contact_flow(event: Any) -> bool:
    return (
        _is_object(event)
        and event.get("Name") == "ContactFlowEvent"
        and _nested_object(event, "Details", "ContactData") is not None
    )


# --- AI / resolver / API auth ---

def is_s3_batch(event: Any) -> TypeGuard[aws_events.S3BatchEvent]:
    return _is_object(event) and _has_string(event, "invocationSchemaVersion") and isinstance(event.get("tasks"), list)


def is_s3_notification(event: Any) -> bool:
    return (
        _is_object(event)
        and event.get("source") == "aws.s3"
        and isinstance(event.get("detail-type"), str)
        and event["detail-type"].strip() != ""
    )


def is_lex(event: Any) -> bool:
    return (
        _is_object(event)
        and _has_string(event, "messageVersion")
        and _is_object(event.get("currentIntent"))
        and _is_object(event.get("bot"))
    )


def is_lex_v2(event: Any) -> bool:
    return (
        _is_object(event)
        and _has_string(event, "messageVersion")
        and _is_object(event.get("bot"))
        and _has_string(event, "sessionId")
    )


def is_appsync_resolver(event: Any) -> TypeGuard[aws_events.AppSyncResolverEvent]:
    if not _is_object(event):
        return False
    info = _nested_object(event, "info")
    return info is not None and _has_string(info, "fieldName") and _is_object(event.get("request"))


def is_amplify_resolver(event: Any) -> bool:
    return (
        _is_object(event)
        and _has_string(event, "typeName")
        and _has_string(event, "fieldName")
        and _is_object(event.get("request"))
    )


def is_api_gateway_authorizer(event: Any) -> bool:
    if not _is_object(event):
        return False
    event_type = event.get("type")
    if event_type not in ("TOKEN", "REQUEST"):
        return False
    return _has_string(event, "methodArn") or _has_string(event, "routeArn")


# --- Infrastructure / device events ---

def is_autoscaling_scale_in(event: Any) -> bool:
    return _is_object(event) and _has_string(event, "AutoScalingGroupARN") and isinstance(event.get("CapacityToTerminate"), list)


def is_cloudformation_custom_resource(event: Any) -> TypeGuard[aws_events.CloudFormationCustomResourceEvent]:
    if not _is_object(event):
        return False
    required_keys = ("RequestType", "ResponseURL", "StackId", "RequestId", "LogicalResourceId", "ResourceType")
    return all(_has_string(event, key) for key in required_keys)


def is_iot_custom_authorizer(event: Any) -> bool:
    return _is_object(event) and _is_object(event.get("protocolData")) and _is_object(event.get("connectionMetadata"))


def is_firehose_transformation(event: Any) -> bool:
    return _is_object(event) and _has_string(event, "invocationId") and isinstance(event.get("records"), list)


def is_secrets_manager_rotation(event: Any) -> TypeGuard[aws_events.SecretsManagerRotationEvent]:
    return _is_object(event) and _has_string(event, "Step") and _has_string(event, "SecretId") and _has_string(event, "ClientRequestToken")


def is_transfer_family_auth(event: Any) -> bool:
    return _is_object(event) and _has_string(event, "username") and _has_string(event, "serverId") and _has_string(event, "sourceIp")
